bus.py

```diff
@@ -0,0 +1,97 @@
+import random
+from utility import distance, firstn, position_on_edge
+
+
+def _road_point_generator(roads):
+    assert len(roads) > 0
+
+    # Length of all roads combined
+    total_length = sum(map(lambda road: road.getLength(), roads))
+
+    while True:
+        # Select a point on the combined stretch of road
+        distance = random.uniform(0, total_length)
+
+        # Find the selected road
+        length_sum = 0.0
+        found_road = False
+        for road in roads:
+            from_node = road.getFromNode().getCoord()
+            to_node = road.getToNode().getCoord()
+
+            if (length_sum + road.getLength()) >= distance:
+                # Compute the exact point on the selected road
+
+                # Distance along the road segment
+                remaining = length_sum - distance
+                pos = position_on_edge(road, remaining)
+
+                yield [
+                        pos[0],
+                        pos[1],
+                        road,
+                        remaining]
+                found_road = True
+                break
+            else:
+                length_sum += road.getLength()
+
+        if not found_road:
+            raise AssertionError("Failed to pick a road. A distance beyound the last road must have been erroneously picked: {} (length sum: {}) (total length: {})".format(distance, length_sum, total_length))
+
+
+def bus_stop_generator(roads, inner_r, outer_r, k=10, seeds=[]):
+    """
+    Bus stop placement using the poisson-disc algorithm
+    """
+    assert inner_r < outer_r
+
+    all_points = list(seeds)
+
+    road_points_gen = _road_point_generator(roads)
+    if len(all_points) <= 0:
+        road = tuple(next(road_points_gen))
+        print(road)
+        yield road
+        all_points.add(road) # Seed point
+
```

Comment:
I am pretty sure something is wrong here `len(all_points) <= 0` is only true if `len(all_points) == 0`, but you are doing `all_points.add(road) # Seed point` in the body, so swapping the `<=` to `>=` would result in an infinite loop.

Reply:
Clarified the intend of the code in 00587e0. I use `len(x) <= y` to check for emptiness and cover the rare case that a length function produces a negative value. I tend to code defensively like that.

Reply:
I see. I get it now. Though typically lengths are not negative 🤔

Reply:
I agree length are **typically** not negative, but this covers faulty code and unknown sentinel values